File: Python3/1001 - 1200/1081 Smallest Subsequence of Distinct Characters.py
// 24ms, 97.67%; 12.7MB, 100.00%
class Solution:
    def smallestSubsequence(self, text: str) -> str:
        res = []
        last, seen = [0] * 26, [False] * 26
        for i, c in enumerate(text):
            last[ord(c) - 97] = i
        for i, c in enumerate(text):
            if seen[ord(c) - 97]: continue
            while res and res[-1] > c and i < last[ord(res[-1]) - 97]:
                seen[ord(res[-1]) - 97] = False
                res.pop()
            seen[ord(c) - 97] = True
            res.append(c)
        return ''.join(res)

File: 1001 - 1200/1081 Smallest Subsequence of Distinct Characters.cpp
// 0ms, 100.00%; 8.6MB, 100.00%
class Solution
{
public:
    string smallestSubsequence(const string &text)
    {
        string res = "";
        int last[26], seen[26] = {}, n = text.size();
        for (int i = 0; i < n; ++i) last[text[i] - 97] = i;
        for (int i = 0; i < n; ++i) {
            if (seen[text[i] - 97]++) continue; // if `res` doesn't contain `text[i]`
            while (!res.empty() && res.back() > text[i] && i < last[res.back() - 97]) // then compare the values
                seen[res.back() - 97] = 0, res.pop_back();
            res.push_back(text[i]);
        }
        return res;
    }
};
